src/history/CheckpointBuilder.cpp

```diff
@@ -0,0 +1,259 @@
+#include "history/CheckpointBuilder.h"
+#include "history/HistoryArchiveManager.h"
+#include "history/HistoryManager.h"
+#include "ledger/LedgerManager.h"
+#include "main/Application.h"
+#include "util/XDRStream.h"
+
+namespace stellar
+{
+void
+CheckpointBuilder::maybeOpen(uint32_t ledgerSeq)
+{
+    ZoneScoped;
+    releaseAssert(mApp.getHistoryArchiveManager().publishEnabled());
+    if (!mOpen)
+    {
+        releaseAssert(!mTxResults);
+        releaseAssert(!mTxs);
+        releaseAssert(!mLedgerHeaders);
+
+        auto checkpoint =
+            mApp.getHistoryManager().checkpointContainingLedger(ledgerSeq);
+        auto res = FileTransferInfo(FileType::HISTORY_FILE_TYPE_RESULTS,
+                                    checkpoint, mApp.getConfig());
+        auto txs = FileTransferInfo(FileType::HISTORY_FILE_TYPE_TRANSACTIONS,
+                                    checkpoint, mApp.getConfig());
+        auto ledger = FileTransferInfo(FileType::HISTORY_FILE_TYPE_LEDGER,
+                                       checkpoint, mApp.getConfig());
+
+        // Open files in append mode
+        mTxResults = std::make_unique<XDROutputFileStream>(
+            mApp.getClock().getIOContext(), /* fsync*/ true);
+        mTxResults->open(res.localPath_nogz_dirty());
+
+        mTxs = std::make_unique<XDROutputFileStream>(
+            mApp.getClock().getIOContext(), /* fsync*/ true);
+        mTxs->open(txs.localPath_nogz_dirty());
+
+        mLedgerHeaders = std::make_unique<XDROutputFileStream>(
+            mApp.getClock().getIOContext(), /* fsync*/ true);
+        mLedgerHeaders->open(ledger.localPath_nogz_dirty());
+        mOpen = true;
+    }
+}
+
+void
+CheckpointBuilder::checkpointComplete(uint32_t checkpoint)
+{
+    ZoneScoped;
+    releaseAssert(mApp.getHistoryArchiveManager().publishEnabled());
+    releaseAssert(
+        mApp.getHistoryManager().isLastLedgerInCheckpoint(checkpoint));
+
+    // This will close and reset the streams
+    mLedgerHeaders.reset();
+    mTxs.reset();
+    mTxResults.reset();
+    mOpen = false;
+
+    auto maybeRename = [&](FileTransferInfo const& ft) {
+        if (fs::exists(ft.localPath_nogz()))
+        {
+            CLOG_INFO(History, "File {} already exists, skipping rename",
```

Comment:
Yes, I want to make sure we can always recover if there are crashes/exceptions thrown. Also, we call `checkpointComplete` on startup, so in case of a graceful restart (nothing to recover), we want to just no-op here.